From ae90cfbd5d9a9d129a7894069ec719a9fc21628e Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:25:46 -0400
Subject: [tools] perf/symbols: Add map_groups__find_ams()

Message-id: <1408462094-14194-178-git-send-email-jolsa@redhat.com>
Patchwork-id: 88183
O-Subject: [PATCH RHEL7.1 BZ1131394 177/325] perf symbols: Add map_groups__find_ams()
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit 4e987712740a3634c19a6fedaf12577b26775dc5
Author: Arnaldo Carvalho de Melo <acme@redhat.com>
Date: Mon Oct 14 13:43:38 2013 +0300

description
===========
Add a function to find a symbol using an ip that might be on a different
map.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/util/map.c b/tools/perf/util/map.c
index 17ee458..9dea404 100644
--- a/tools/perf/util/map.c
+++ b/tools/perf/util/map.c
@@ -371,6 +371,23 @@ struct symbol *map_groups__find_symbol_by_name(struct map_groups *mg,
  return NULL;
 }
 
+int map_groups__find_ams(struct addr_map_symbol *ams, symbol_filter_t filter)
+{
+ if (ams->addr < ams->map->start || ams->addr > ams->map->end) {
+  if (ams->map->groups == NULL)
+   return -1;
+  ams->map = map_groups__find(ams->map->groups, ams->map->type,
+         ams->addr);
+  if (ams->map == NULL)
+   return -1;
+ }
+
+ ams->al_addr = ams->map->map_ip(ams->map, ams->addr);
+ ams->sym = map__find_symbol(ams->map, ams->al_addr, filter);
+
+ return ams->sym ? 0 : -1;
+}
+
 size_t __map_groups__fprintf_maps(struct map_groups *mg,
       enum map_type type, int verbose, FILE *fp)
 {
diff --git a/tools/perf/util/map.h b/tools/perf/util/map.h
index 4886ca2..0359b4a 100644
--- a/tools/perf/util/map.h
+++ b/tools/perf/util/map.h
@@ -167,6 +167,10 @@ struct symbol *map_groups__find_symbol_by_name(struct map_groups *mg,
             struct map **mapp,
             symbol_filter_t filter);
 
+struct addr_map_symbol;
+
+int map_groups__find_ams(struct addr_map_symbol *ams, symbol_filter_t filter);
+
 static inline
 struct symbol *map_groups__find_function_by_name(struct map_groups *mg,
        const char *name, struct map **mapp,
-- 
1.7.1